### ISSUE TITLE
6.1.1 oneof validate error

### ISSUE THREAD

**Issue by nick121212** [ajv-validator/ajv#692]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**

6.1.1

**Ajv options object**

```javascript

new ajv({
    allErrors: true,
    jsonPointers: true,
    useDefaults: true,
    format: "full",
    $data: true,
    errorDataPath: "property",
    removeAdditional: true,
})

```

**JSON Schema**

```javascript

{
    type: "object",
    $id: "design",
    required: ["name"],
    properties: {
        name: {
            type: "string"
        },
        appType: {
            oneOf: [{
                $id: "design-app-type-1",
                type: "object",
                properties: {
                    a: {
                        type: "string",
                        title: "test-oneof-2"
                    }
                }
            }, {
                $id: "design-app-type-2",
                type: "object",
                properties: {
                    b: {
                        type: "string",
                        title: "test-oneof-2"
                    }
                }
            }]
        }
    }
}

```

**Sample data**

```javascript

{
    name: "nick",
    appType: {
        a: "dkfkald"
    }
}

```

**Your code**

```javascript


```

**Validation result, data AFTER validation, error messages**

```
[{
dataPath:"/appType",
keyword:"oneOf",
message:"should match exactly one schema in oneOf",
schemaPath:"#/properties/appType/oneOf"
}]

```

**What results did you expect?**

validate should be true

**Are you going to resolve the issue?**

**Comment by epoberezkin** [comment]
oneOf means that only one sub-schema is valid, in your case - both are valid.

**Comment by nick121212** [comment]
thanks.

**Comment by jscharett** [comment]
I'm not sure I understand this.  I've run into this issue and can validate it with other validators, but not this one.  You say they are both valid?  Should it not match the first one but not the second, thus being valid?

In my case, if I change the order of the oneOf, it affects the validation. I'm not sure how to validate such a case with your tool.

**Comment by epoberezkin** [comment]
TLDR: please read the JSON schema spec and/or KEYWORDS.md file in this repo.

They are both valid because neither schema requires or prohibits the presence of any property. Specifically, the data is valid against the first schema, because “a” is present and matches the schema for “a”. The data is valid against the second schema, because there is no property “b”.

That’s a very common misconception to think that JSON schema describes the structure the data should have to be valid, while it just defines conditions that make data invalid if they are not met. In this way “properties” keyword neither requires any property to be present, nor prohibits the presence of any properties, it just defines conditions that a property should satisfy IN CASE it is present. Also, it does not require any type of the data, any non-object will pass validation with “properties” keyword (see #731 for such issue). When such mistakes happen in the context of oneOf, they are easily noticeable, as the validation correctly,  but unexpectedly, fails. But without oneOf such mistakes simply lead to false positive results, that is probably one of the biggest problems of JSON schema in general, as many people start using it relying on their expectations about how it should work without looking at the spec/tutorials... /rant :)

If you have a specific example that you believe is not working according to the spec, please post it here but it would help if you checked the spec as well.

**Comment by epoberezkin** [comment]
As a note about data passing validation with some other validators. The fact that the validation results differ between validators means just that - that the results are different. It obviously means that one of (oneOf:) validators is violating the spec (unless you’re in the grey zone of format validation). But it does not mean that the one that fails is incorrect. From my experience, the opposite is much more likely - i.e. the one that passes is incorrect, because the whole purpose of the validation is to recognise certain data as invalid.

The simplest possible validator would just always return “valid” and it is likely to be correct most of the time (at least, server-side), but such validator would be useless.

**Comment by jyotivit** [comment]
oneOf means that only one sub-schema is valid, in your case - both are valid. it is the correct as i was stuck for 4 day before i got this post that resolved my issue  thanks

**Comment by vkurlyan** [comment]
You might add `additionalProperties: false` to both objects in oneOf, so they won't be valid simultaneously.
